File: Theory/DublyLinkedList.py
class Node:
    def __init__(self, data, nextNode, prevNode):
        self.data = data
        self.nexNode = nextNode
        self.prevNode = prevNode

class DoublyLinkedList:
    def __init__(self):
        self.listSize = 0
        self.head = None
        self.tail = None
        
    def addFirst(self, data):
        if (self.listSize == 0):
            node = Node(data, None, None)
            self.head = node
            self.tail = node
        else:
            node = Node(data, self.head, None)
            self.head.prevNode = node
            self.head = node
        self.listSize += 1
        
    def addLast(self, data):
        if (self.listSize == 0):
            self.addFirst(data)
        else:
            node = Node(data, None, self.tail)
            self.tail.nexNode = node
            self.tail = node
            self.listSize += 1
        
    def getNodeFromFront(self, idx):
        node = self.head
        for i in range(0, idx):
            node = node.nexNode
        return node.data
    
    def getNodeFromEnd(self, idx):
        node = self.tail
        for i in range(self.listSize - 1, idx, -1):
            node = node.prevNode
        return node.data

    def getNode(self, node):
        return node.data
        
    def search(self, value):
        node = self.head
        for i in range(0, self.listSize):
            if (node.data == value):
                return node
            node = node.nexNode
        return node
        
    def remove(self, node):
        if (self.listSize == 1):
            self.head == None;
            self.tail == None;
        else:
            if (node == self.head):
                self.head = self.head.nexNode
                self.head.prevNode = None
            elif (node == self.tail):
                self.tail = self.tail.prevNode
                self.tail.nexNode = None
            else:
                node.prevNode.nexNode = node.nexNode
                node.nexNode.prevNode = node.prevNode
                node.prevNode = None
                node.nexNode = None
        self.listSize -= 1       
dbl = DoublyLinkedList();
dbl.addFirst(1)
dbl.addFirst(2)
dbl.addLast(5)
dbl.addLast(7)
dbl.addLast(8)

for i in range(0, dbl.listSize):
    print(dbl.getNodeFromEnd(i))

print("=============")

dbl.remove(dbl.search(5))


for i in range(0, dbl.listSize):
    print(dbl.getNodeFromEnd(i))

File: Theory/DublyLinkedList/DublyLinkedList.java
class DoublyLinkedList {
    int listSize = 0;
    Node head = null;
    Node tail = null;

    class Node {
        Node previous = null;
        Node next = null;
        int value = 0;

        private Node(Node previousNode, Node nextNode, int value) {
            this.previous = previousNode;
            this.next = nextNode;
            this.value = value;
        }

        private void setPreviousNode(Node previousNode) {
            this.previous = previousNode;
        }

        private void setNextNode(Node nextNode) {
            this.next = nextNode;
        }

        private Node getPreviousNode() {
            return this.previous;
        }

        private Node getNextNode() {
            return this.next;
        }

        private int getValue() {
            return value;
        }
    }

    public void addLast(int value) {
        if (listSize == 0) {
            addFirst(value);
        } else {
            Node newNode = new Node(tail, null, value);
            tail.setNextNode(newNode);
            tail = newNode;
            listSize++;
        }
    }

    public void addFirst(int value) {
        if (listSize == 0) {
            Node newNode = new Node(null, null, value);
            head = newNode;
            tail = newNode;
            listSize++;
        } else {
            Node newNode = new Node(null, head, value);
            head.setPreviousNode(newNode);
            head = newNode;
            listSize++;
        }
    }

    private Node getNodeFromFront(int index) {
        if (index > listSize) {
            return null;
        }

        Node node = head;

        for (int i = 0; i < index; i++) {
            node = node.getNextNode();
        }
        return node;
    }

    private Node getNodeFromEnd(int index) {
        if (index > listSize) {
            return null;
        }

        Node node = tail;

        for (int i = listSize - 1; i > index; i--) {
            node = node.getPreviousNode();
        }

        return node;
    }

    public Node getNode(int index) {
        if (index <= listSize / 2) {
            Node node = getNodeFromFront(index);
            return node;
        } else {
            Node node = getNodeFromEnd(index);
            return node;
        }
    }

    public static int getNodeValue(Node node) {
        return node.getValue();
    }

    public Node search(int value) {
        for (int i = 0; i < listSize; i++) {
            if (value == getNodeValue(getNode(i))) {
                return getNode(i);
            }
        }
        return null;
    }

    public void remove(Node node) {
        if (listSize == 1) {
            head = null;
            tail = null;
        } else {
            if (node == head) {
                head.getNextNode().setPreviousNode(null);
                head = head.getNextNode();
            } else if (node == tail) {
                tail.getPreviousNode().setNextNode(null);
                tail = tail.getPreviousNode();
            } else {
                node.getPreviousNode().setNextNode(node.getNextNode());
                node.getNextNode().setPreviousNode(node.getPreviousNode());
                node.setPreviousNode(null);
                node.setNextNode(null);
            }
        }
        listSize--;
    }

}

class Main {
    public static void main(String[] args) {
        DoublyLinkedList dbl = new DoublyLinkedList();
        dbl.addFirst(1);
        dbl.addFirst(2);
        dbl.addLast(3);
        dbl.addLast(5);
        dbl.addLast(7);
        dbl.addLast(9);

        for (int i = 0; i < dbl.listSize; i++) {
            System.out.println(DoublyLinkedList.getNodeValue(dbl.getNode(i)));
        }

        System.out.println("----------------------");

        dbl.remove(dbl.search(2));

        for (int i = 0; i < dbl.listSize; i++) {
            System.out.println(DoublyLinkedList.getNodeValue(dbl.getNode(i)));
        }
    }
}